Rebrand Diary View to Plan View

paulj49457 opened the pull request:
To support the new planning chart(s) the Diary View has been updated to Plan View.

This is an initial start to the full Plan View, the following have been updated:

- Class name DiaryView -> PlanView
- Class name DiarySidebar -> PlanSideBar
- Added "Plan" button to Edit Menu
- Diary configuration changed to Plan configuration names
- New plan-perspective.xml configuration file
- New default "baked in" plan-perspective.xml configuration file with single PlanningCalendarWindow chart
- "Legacy" perspective name has been restricted to be applied only for the old -layout.xml files
- Tested PlanningCalendarWindow export and import
- Tested NavigationModel, see notes below

The following have been noticed or need to be addressed:

i) I have assumed that re-using the old diary 0x04 view type src/Gui/GcWindowRegistry.h() is ok ?

ii) The navigation undo actions can cause a crash!  After changing the timeframe in the trends view, which then caused the context->currentSeason() to return nullptr. I have added protection for this, please search for PLAN_VIEW_TODO

iii) Should the PlanView and TrendsView have independent date ranges (search for PLAN_VIEW_TODO) ? At the moment they use the same one, which results in some interesting behaviour, for example, if dates only in 2024 are set in trends, when switching to the Plan View will not allow the new planning calendar to display dates in 2025, it is restricted to 2024 dates. Now the planning chart is in its own view rather than part of the trends view, it may need its only independent date range.

Please let me know if anything is wrong, could be improved or needs to be update prior to merging

amtriathlon commented:
> The following have been noticed or need to be addressed:
> 
> i) I have assumed that re-using the old diary 0x04 view type src/Gui/GcWindowRegistry.h() is ok ?

Yes, but for now I would let all charts enabled for DIARY_VIEW enabled for PLAN_VIEW, besides the new Calendar chart.
 
> ii) The navigation undo actions can cause a crash! After changing the timeframe in the trends view, which then caused the context->currentSeason() to return nullptr. I have added protection for this, please search for PLAN_VIEW_TODO

Right, usually there is a Season/DateRange (and a Ride) selected, but the code should not crash when they are not.
 
> iii) Should the PlanView and TrendsView have independent date ranges (search for PLAN_VIEW_TODO) ? At the moment they use the same one, which results in some interesting behaviour, for example, if dates only in 2024 are set in trends, when switching to the Plan View will not allow the new planning calendar to display dates in 2025, it is restricted to 2024 dates. Now the planning chart is in its own view rather than part of the trends view, it may need its only independent date range.

No, I don't think so, that would require extensive changes to the APIs (DataFilter/R/Python). Lets think about other solution for this.

paulj49457 commented:
> Thanks, Paul, it looks ok for me.

PR updated with the changes you suggested:
> 
> > The following have been noticed or need to be addressed:
> > i) I have assumed that re-using the old diary 0x04 view type src/Gui/GcWindowRegistry.h() is ok ?
> 
> Yes, but for now I would let all charts enabled for DIARY_VIEW enabled for PLAN_VIEW, besides the new Calendar chart.

I have re-enabled all the original Diary charts to be available under Plan view.
I have also made the new Calendar chart available in both trends and plan, otherwise I'd also need to update the baked in plan perspective I have created. This will allow testing of the new Calendar chart, as we work out the issues of moving it from trends to plan.  (Let me know if the new Calendar chart must only be in Diary view)
> 
> > ii) The navigation undo actions can cause a crash! After changing the timeframe in the trends view, which then caused the context->currentSeason() to return nullptr. I have added protection for this, please search for PLAN_VIEW_TODO
> 
> Right, usually there is a Season/DateRange (and a Ride) selected, but the code should not crash when they are not.

 I have added protection for this, and removed the todo comment.
> 
> > iii) Should the PlanView and TrendsView have independent date ranges (search for PLAN_VIEW_TODO) ? At the moment they use the same one, which results in some interesting behaviour, for example, if dates only in 2024 are set in trends, when switching to the Plan View will not allow the new planning calendar to display dates in 2025, it is restricted to 2024 dates. Now the planning chart is in its own view rather than part of the trends view, it may need its only independent date range.
> 
> No, I don't think so, that would require extensive changes to the APIs (DataFilter/R/Python). Lets think about other solution for this.

 I was just trying to raise the issue which I don't have enough knowledge to resolve, I have removed the todo comment.

paulj49457 commented:
I have updated this PR to correct the new code (VIEW_DIARY to VIEW_PLAN) added by #4698

paulj49457 commented:
@amtriathlon Please see latest commit for a fix for the Metric Trends and User Charts don't respond to Date Range changes when they are in Plan View: https://github.com/GoldenCheetah/GoldenCheetah/pull/4690/commits/4d61707f4055ad36295f2a0a55dff4f757a60b57
